### PR TITLE
WASM Native FileOpenPicker

### LINKED ISSUE #5323: [WASM] Native FileOpenPicker
## What would you like to be added:

Represents the file system access based version of #4028

## For which Platform:

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): part of #4028, closes #5323

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

WASM FileOpenPicker not available

## What is the new behavior?

WASM FileOpenPicker using native file system APIs.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/ts/Windows/Storage/Pickers/FileOpenPicker.Native.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿namespace Windows.Storage.Pickers {
+
+	export class FileOpenPicker {
+		public static async pickFilesAsync(multiple: boolean, showAllEntry: boolean, fileTypes: any): Promise<string> {
+
+			if (!showOpenFilePicker) {
+				return "";
+			}
+
+			var options: OpenFilePickerOptions  = {
```
Comment (carldebilly) [inline src/Uno.UI/ts/Windows/Storage/Pickers/FileOpenPicker.Native.ts:10]: I agree with Codacy....

**Review by carldebilly — COMMENTED** [review]
Is there a way to unit test this?

File: src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+			var multipleParameter = multiple ? "true" : "false";
+			var fileTypeMapParameter = BuildFileTypesMap();
+
+			var returnValue = await WebAssemblyRuntime.InvokeAsync($"{JsType}.pickFilesAsync({multipleParameter},{showAllEntryParameter},{fileTypeMapParameter})");
+			var files = returnValue.Split(new string[] { FileSeparator }, StringSplitOptions.RemoveEmptyEntries);
+			var results = new List<StorageFile>();
+			foreach (var file in files)
+			{
+				var fileInfoParts = file.Split(FileInfoSeparator);
+				Guid.TryParse(fileInfoParts[0], out var guid);
```
Comment (jeromelaban) [inline src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs:42]: What if this fails ?

File: src/Uno.UWP/Storage/StorageFile.native.wasm.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+using Windows.Storage.FileProperties;
+using Windows.Storage.Streams;
+using SystemPath = global::System.IO.Path;
+
+namespace Windows.Storage
+{
+	public partial class StorageFile
+	{
+		internal static StorageFile GetFileFromNativePathAsync(Guid guid, string name, string contentType) => new StorageFile(new NativeFileSystem(guid, name, contentType));
```
Comment (jeromelaban) [inline src/Uno.UWP/Storage/StorageFile.native.wasm.cs:13]: Why did you add `Async` there ?

File: src/Uno.UWP/Storage/StorageFile.native.wasm.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+		private sealed class NativeFileSystem : ImplementationBase
```
Comment (jeromelaban) [inline src/Uno.UWP/Storage/StorageFile.native.wasm.cs:15]: What about this (same for the javascript one) ? 

```suggestion
		private sealed class NativeStorageFile : ImplementationBase
```

**Review by jeromelaban — COMMENTED** [review]
I agree with most of what the robot said :)

**Comment by MartinZikmund** [comment]
> Is there a way to unit test this?

Unfortunately not testable, but I will try to add tests for the file operations (reading/writing) which will be in the next PR

**Comment by jeromelaban** [comment]
Can you tag the manual tests using the new attribute? https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#adding-a-manual-test-sample

**Comment by MartinZikmund** [comment]
> Can you tag the manual tests using the new attribute? https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#adding-a-manual-test-sample
> 
> >  GitHub**[unoplatform/uno](https://github.com/unoplatform/uno)**Build Mobile, Desktop and WebAssembly apps with C# and XAML. Today. Open source and professionally supported. - unoplatform/uno

Done ✅

File: src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+				}
+				var name = fileInfoParts[1];
+				var contentType = fileInfoParts[2];
+				var storageFile = StorageFile.GetFileFromNativePath(guid, name, contentType);
+				results.Add(storageFile);
+			}
+			return new FilePickerSelectedFilesArray(results.ToArray());
+		}
+
+		private string BuildFileTypesMap()
```
Comment (carldebilly) [inline src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs:54]: Maybe this should be cached?

**Review by carldebilly — APPROVED** [review]

File: src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs (hunk 5 above)
Comment (MartinZikmund) [reply src/Uno.UWP/Storage/Pickers/FileOpenPicker.wasm.cs:54]: The issue is the file types may change any time, so I would have to add some event to the `FileTypeFilter` to notify about this or something similar. It should not matter too much though, the number of mime extensions is low, so it will not have noticeable impact.

File: src/Uno.UWP/WinRTFeatureConfiguration.cs [hunk 6]
```diff
@@ -64,5 +64,13 @@ public static class NetworkInformation
 		{
 			public static string ReachabilityHostname { get; set; } = "www.example.com";
 		}
+
+		public static class FileTypes
+		{
+			/// <summary>
+			/// Allows definition of custom file type MIME mappings.
+			/// </summary>
+			public static Dictionary<string, string> FileTypeToMimeMapping { get; } = new Dictionary<string, string>();
```
Comment (jeromelaban) [inline src/Uno.UWP/WinRTFeatureConfiguration.cs:73]: Should this be documented ?

Comment (MartinZikmund) [reply src/Uno.UWP/WinRTFeatureConfiguration.cs:73]: I am planning to document when I do some full docs for all pickers (it will be used on other platforms too)

**Review by jeromelaban — APPROVED** [review]
